### DISCUSSION TITLE
Bad sensitivity for high values of alphaPorosity

### DISCUSSION THREAD

**Discussion by schewef** [mdolab/dafoam#208] [General]
Hello,

I did some sensitvity evaluation considering the alphaPorosity value in simpleTFoam.

Comparing the adjoint sensitivities against finite difference sensitivities I noticed that they match perfectly for low values of alphaPorosity. The solution of the linear system of equations unsing the KSP solver works super fast.

If I have a fully filled model e.g. with all alphaPorosity values at 1e+6 the KSP solver convergence is very bad and also the adjoint sensitivities do not match the finite difference sensitivities even though the KSP Residual converges after ~20000 iterations.

The phenomenon seems to be very dependent on the optimization case. The pitzDaily case does not seem to have that big issues.
Did you ever experienced such issues? Do you have some hint how to trubleshoot it?

My guess is that the matrix system gets badly conditioned if all design alphaPorosity values are at 1e+6. Because of the resulting bad convergence the numerical error might sum up over the massive amount of KSP iterations.
Is there a possibility to normalize the used vectors / matrices or does dafoam already does this internally?

I already tried to use all other preconditioners availabe in petsc but that does not seem to solve the problem.
I also tried to change the "pcFillLevel" and "jacMatReOrdering" without success.

I would be very gratefull for any hints.

Regards,
schewef

**Comment by friedenhe** [comment]
Yes, what you observed is probably the cause of the poor convergence of the adjoint. We did not scale alphaPorosity when computing the KSP precondition matrix, which may result in bad conditions for the matrix. We have not tested the adjoint with large alphaPorosity and do not have a good solution at this moment. Let us know if you manage to fix the issue.

**Reply by schewef** [reply]
Thanks for your reply. I invested some time and was able to get better convergence by scaling the residual vector.

Anyhow there seems to be another issue.

I printed the dFdW derivatives and noticed that it is always zero, whatever the design variables are and whatever the objective function is. I checked it for multiple testcases.

The same phenomenon occurs for dFdField.

Does this make any sense?

I attached the pitzDaily testcase with exported dFdW and dFdField values as:
dFdField_PL.dat / dFdW_PL.dat
[pitzDaily.zip](https://github.com/mdolab/dafoam/files/7651289/pitzDaily.zip)

It was executed inside the docker image just with one modification to export dFdField with:
mpirun -np 4 python runScript.py --task=runAdjoint 2>&1 | tee logOpt.txt

**Comment by friedenhe** [comment]
I suspect you are not writing the Jacobians in the right way. DAFoam has an option to output these Jacobians. Check this option:
https://github.com/mdolab/dafoam/blob/e5c2b180527eda2e192a253d0e19aa5b856726d6/dafoam/pyDAFoam.py#L478

**Reply by schewef** [reply]
Thanks for the reply!

Using this option is exactly what I did to export dFdW. 
The output is still zero.

dFdField is not supported with this function therefore I just added one line of code to export it after calculation via

[self.solverAD.calcdFdFieldAD(self.xvVec, self.wVec, objFuncName.encode(), designVarName.encode(), dFdField)](https://github.com/mdolab/dafoam/blob/e5c2b180527eda2e192a253d0e19aa5b856726d6/dafoam/pyDAFoam.py#L2287)
self.writePetscVecMat("dFdField_%s" % objFuncName, dFdField, "ASCII")

Everything else is exacly like in the original docker version, tested with the original pitzDaily tutorial.

I suspect that the AD recording somehow does not work but I was not able to find out why this is the case.

**Comment by friedenhe** [comment]
dFdField should be zero because the pressure loss (PL) obj does not explicitly depend on the field design variable. For dFdW, it should not be all zeros. Note that dFdW should have just a few non-zero elements, most of dFdW should be zeros.

**Reply by schewef** [reply]
Oh you are absolutely right. Sorry for that stupid question.
dFdW in fact has some non-zero values. 

I was confused because dFdField was all zero, but if this is correct it all seems to work correctly. Then the issue has to be on my side.
Thank you very much for your helpfull reply!

**Reply by qwefhajk** [reply]
I tried to compare the accuracy of the adjoint sensitivities with the finite difference method regarding the SimpleFoam in pitzDaily case, and the results are not satisfactory (the relative errors are mainly larger than 5% and some sample points might reach up to 50%). May I ask what's the results of your comparison experiments and how did you conduct the finite difference method in pitzDaily case?

**Reply by schewef** [reply]
To extract the finite difference sensitivities you just have to figure out a reasonable step size. Afterwards you can choose random elements modify the specific value for alphaPorosity for that element by the step size and call the primal solver two times (at initial design and at design with alphaPorosity modified by step size). Form the resulting objective values you can calculate e.g. the "forward-difference".

I guess DAFoam already has a methodology implemented to calculate the FD sensitivities but I wrote my own routine to extract them.

I tested the sensitivities at two states:
1. unfilled design (all alphaPorosity values = 0)
2. filled design (all alphaPorosity values = 1e+4)

For the unfilled design the adjoint and FD sensitivities match quite well in absolute terms. You are right some relative errors can get up to 50% or higher but the absolute difference is still small, so thats fine for me.

For the filled design the absolute difference is much higher, and even gets worse for other testcases cases apart from pitzDaily. I am currently not sure what is the reason for that effect.

**Comment by sodabryant** [comment]
@schewef , sorry to disturb you, it seems that you have sucessully modify the topology optimization solver in DAFoam. May I ask which optimizer you are using? I can't find any optimizer to replace the SNOPT, you know, which is is a commercial package. So could you please share any optimizer packages or methods you have tried? My email address is lixinlei@imech.ac.cn. I would be grateful for any advises.
Happy new year and best regards!
